Basic matrix operation using numpy: \

add() − add elements of two matrices.

subtract() − subtract elements of two matrices.

divide() − divide elements of two matrices; This is not matrix division

multiply() − multiply elements of two matrices.

dot() − It performs matrix multiplication, not element wise multiplication; notice difference between multiply and dot; Don't be confused with inner product of vectors

sqrt() − square root of each element of matrix.

sum(x,axis) − add to all the elements in matrix. Second argument is optional, it is used when we want to compute the column sum if axis is 0 and row sum if axis is 1.

“T” − It performs transpose of the specified matrix.

linalg.det() - calculates the determinant of a matrix

Rank

Trace

Inverse and pseudo-inverse

Flatten

Eigenvlues and Eigenvectors

In [2]:
import numpy as np

In [5]:
A = [[1, 2], [3, 4]]
A, type(A)

([[1, 2], [3, 4]], list)

Here A is a list. It needs to be in array for matrix operations.
Convert list A to an array A

In [3]:
A = np.array([[1, 2], [3, 4]])
A, type(A)

(array([[1, 2],
        [3, 4]]), numpy.ndarray)

Let's create another matrix B and C.

In [14]:
B = np.array([[5, 6], [7, 8]])
B

array([[5, 6],
       [7, 8]])

In [17]:
C = np.array([[1, 2, 3], [4, 5, 6]])
C

array([[1, 2, 3],
       [4, 5, 6]])

Dtermine the shape of a matrix A, B, and C

In [25]:
np.shape(A), np.shape(B), np.shape(C)

((2, 2), (2, 2), (2, 3))

Adding two matrix with same shape. A + B

In [13]:
add_matrix = np.add(A, B)
add_matrix

array([[ 6,  8],
       [10, 12]])

Subtracting two matrix with same shape. A - B


In [16]:
sub_matrix1 = np.subtract(A, B)
sub_matrix1

array([[-4, -4],
       [-4, -4]])

Subtracting two matrix with same shape. B - A


In [17]:
sub_matrix2 = np.subtract(B, A)
sub_matrix2

array([[4, 4],
       [4, 4]])

If you subtract matrix from itself, you get null matrix. See in this example B - B

In [18]:
sub_matrix3 = np.subtract(B, B)
sub_matrix3

array([[0, 0],
       [0, 0]])

np.multiply doesn't give the product of two matrices. instead it is used to multiply corresponding elrmrnts.

In [21]:
mul_matrix1 = np.multiply(A, B)
mul_matrix1

array([[ 5, 12],
       [21, 32]])

If dimensions of matrices don;t match, you cannot multiply. See below

In [23]:
mul_matrix2 = np.multiply(A, C)
mul_matrix2

ValueError: ignored

Two matrices are multiplied below

In [26]:
mul_matrix3 = np.dot(A, B)
mul_matrix3

array([[19, 22],
       [43, 50]])

Calculate the determinant of a matrix

In [28]:
det_A = np.linalg.det(A)
np.round(det_A)

-2.0

Transpose of a matrix A

In [38]:
A.T

array([[1, 3],
       [2, 4]])

The rank of a matrix is the dimensions of the vector space spanned (generated) by its columns or rows.

In [8]:
np.linalg.matrix_rank(A)

2

The trace of a matrix is the sum of its diagonal elements.

In [9]:
np.trace(A)

5

If A.B = I (Iidentity matrix - matrix with diangonal elements 1 and non diangonal elements zero), then B is inverse of A.

In [12]:
D = np.linalg.inv(A)
D

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

Generally, there is no inverse of a singular matrix (whose determinant is zero). But for computational problems, we need to calculate pseudo inverse to avoid error.

In [21]:
D = np.array([[2,8], [1, 4]])
np.linalg.pinv(D)

array([[0.02352941, 0.01176471],
       [0.09411765, 0.04705882]])

In [13]:
np.dot(A, D)

array([[1.0000000e+00, 0.0000000e+00],
       [8.8817842e-16, 1.0000000e+00]])

A.flatten() will convert nXm array into one dimensional array.

In [18]:
C.flatten()

array([1, 2, 3, 4, 5, 6])

Let T be an n x n matrix. A scalar λ is called an eigenvalue of T if there is a non-zero vector v satisfying the following equation.

                      Tv = λv

where v is eigenvector of T corresponding of λ.



In [4]:
w, v = np.linalg.eig(A)
w, v

(array([-0.37228132,  5.37228132]), array([[-0.82456484, -0.41597356],
        [ 0.56576746, -0.90937671]]))

The eigenvalue is w

In [5]:
w

array([-0.37228132,  5.37228132])

The eigenvector is v

In [6]:
v

array([[-0.82456484, -0.41597356],
       [ 0.56576746, -0.90937671]])

In [7]:
sum(w)

5.0

In [8]:
np.trace(A)

5

The sum of eigenvalues (-0.37 + 5.37 = 5) is equal to the trace (1 + 4 = 5) of the same matrix! The product of the eigenvalues (-0.37*5.37=-2) is equal to the determinant (-2) of the same matrix!

Eigenvalues and eigenvectors are extremely useful in the Principal Component Analysis (PCA). In PCA, the eigenvectors of the correlation or covariance matrix represent the principal components (the directions of maximum variance) and the corresponding eigenvalues represent the amount of variation explained by each principal component.